### PR TITLE
[react] [react-dom] Add support for rendering an array of elements

### PR DESCRIPTION
**Pull request by cynecx**
- [X] Use a meaningful title for the pull request. Include the name of the package modified.
- [X] Test the change in your own code. (Compile and run.)
- [X] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [X] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [X] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).
- [X] Provide a URL to documentation or source code which provides context for the suggested changes:
    -  Callbacks don't have any parameters (See commitCallbacks): https://github.com/facebook/react/blob/master/src/renderers/shared/fiber/ReactFiberUpdateQueue.js
- [X] Increase the version number in the header if appropriate.
- [X] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`.

Changes:

- Added more overloads to `ReactDOM::render` to support rendering an array of elements
    - Since typescript doesn't support variadic types, the array overloads use `any` instead of a deduced type
- The callback function signature has been changed because react v16 doesn't actually pass any parameters anymore
- Bumped the supported react-dom to v16

I am not quite sure if these changes are correct, so please be gentle on reviewing these changes ;)

### CODE REVIEW

**Comment by cynecx** [comment]
Uhmm, I guess something is clearly wrong. Checking it out.

**Comment by cynecx** [comment]
It seems that this patch doesn't affect any of the occurring errors as I can reproduce them even without this patch. Would be great if someone could review this. Thank You.

File: types/react-dom/index.d.ts
Comment (vsiao) [inline types/react-dom/index.d.ts]: can you link to any documentation of this API? I didn't know that it was possible to pass an array into `ReactDOM.render`.

**Review by vsiao — CHANGES_REQUESTED** [review]
I think the `react-dom` changes are incorrect. Core `react` changes seem fine, but perhaps missing changes to SFCs?

File: types/react-dom/index.d.ts
Comment (cynecx) [reply types/react-dom/index.d.ts]: It's not yet documented but the with React Fiber (v16) you can pass an array of elements to the render function.

See this test case: https://codesandbox.io/s/y310mjkp5v

The test case is using a beta release of react/react-dom (16.0.0-beta.5).

Comment (cynecx) [reply types/react-dom/index.d.ts]: Just as a code reference:

https://github.com/facebook/react/blob/master/src/renderers/shared/fiber/ReactChildFiber.js#L565

The parameter `newChild` is basically the `element` passed to the render function. This line in `createChild` checks whether the `element` parameter, which has been passed down to this function, is an array. If so it deals with it respectively which indicates that React-Fiber (v16) supports passing an array of elements to the render function (react-dom).

Comment (vsiao) [reply types/react-dom/index.d.ts]: Thanks. Can you add type parameters for props, as in the other overloads?

File: types/react-dom/index.d.ts
Comment (vsiao) [inline types/react-dom/index.d.ts]: should this be an array return value?

File: types/react-dom/index.d.ts
Comment (cynecx) [reply types/react-dom/index.d.ts]: Excuse me, but how would that work? Since we don't have something like "variadic parameter types" in typescript. Am I missing something?

File: types/react-dom/index.d.ts
Comment (cynecx) [reply types/react-dom/index.d.ts]: I don't think so. It seems that React-Fiber does not return the whole array back which is somewhat strange. According to my tests, it only returns the first element of the array.

**Comment by cynecx** [comment]
It seems that there is in issue in `react-router/test/InheritingRoute.tsx`. The problem here is that the return type of the render function (`JSX.Element | JSX.Element[]`) is not compatible to the paramter type of the cloneElement function signature as it only accept `JSX.Element`.

I am not sure how to fix this issue. We can either assume that the render function in the class Route from react-router always return an JSX.Element or we can modify the test.

We could modify the Route class like the following:

```
export class Route<T extends RouteProps = RouteProps> extends React.Component<T> {
  render(): JSX.Element | null | false;
}
```

Should this go in a new PR?

**Comment by richseviora** [comment]
@cynecx I'm thinking it's probably best to update the broken tests (`react-router` and other types as necessary) and then flag their contributors about the change for their approval.

**Comment by cynecx** [comment]
Rebased on master. I am working on fixing the tests now.

**Comment by cynecx** [comment]
Seems good now. There is one error which I am certain that this PR didn't affect:

```
Error in redux-form v6
Command failed: node /home/travis/build/DefinitelyTyped/DefinitelyTyped/node_modules/dtslint/bin/index.js --noLint
Unexpected compiler option noImplicitReturns
```

**Comment by cynecx** [comment]
Flagging some users for approval: @prakarshpandey, @andy-ms, @mtpc, @tkrotoff, @MartynasZilinskas

**Comment by circlingthesun** [comment]
Any update on this?

**Comment by cynecx** [comment]
@circlingthesun Well, I am still currently waiting for approval :)

**Comment by andy-ms** [comment]
Hi @cynecx, tests should be fixed if you merge in the latest master.

**Comment by cynecx** [comment]
@andy-ms Okay. I am not sure how this PR caused that last one but I'll fix it :)

**Comment by mhegazy** [comment]
@cynecx, can you address the failed tests.

**Comment by cynecx** [comment]
@mhegazy It seems to me that that last issue with the linter is a false-positive, hence I've opened https://github.com/Microsoft/dtslint/issues/72. But it seems that the development of dtslint has been stagnating. I tried to workaround this issue but it didn't work well. I might give this another try later.

**Comment by circlingthesun** [comment]
@cynecx, I really appreciate you taking this on. :+1:

**Comment by cynecx** [comment]
:+1:

**Comment by andy-ms** [comment]
@cynecx You may need to merge in the latest master branch.

**Comment by mhegazy** [comment]
@vsiao any more comments?

**Comment by cynecx** [comment]
I've rebased the branch against master.

File: types/react-dom/index.d.ts
Comment (Kovensky) [reply types/react-dom/index.d.ts]: It could be `Array<DOMElement<DOMAttributes<any>, any>>>` for the lack of better specificity, to mirror the generic overload above that has `DOMAttributes<T>` as the props.

File: types/react-dom/index.d.ts
Comment (Kovensky) [reply types/react-dom/index.d.ts]: Returning something from `render` is itself deprecated. Function `ref`s should be used.

I'd be in favor of declaring this as `void` even though it's technically _de facto_ wrong.

Comment (cynecx) [reply types/react-dom/index.d.ts]: I would leave this as it is for now. When it's finally changed we can reflect the changes afterwards. It may be deprecated but some users might still use this in their code. I guess it's better this way.

File: types/react-dom/index.d.ts
Comment (cynecx) [reply types/react-dom/index.d.ts]: I agree. I've commited the change you suggested. Thanks.

File: types/react-dom/index.d.ts
Comment (MartynasZilinskas) [reply types/react-dom/index.d.ts]: At least adding a comment with a deprecated tag should be enough in my opinion.

Comment (cynecx) [reply types/react-dom/index.d.ts]: Where would you like to add them? Marking the whole render function as deprecated doesn't feel right (Using jsdoc's @deprecated). Or do you think a simple comment would be sufficient?

Comment (MartynasZilinskas) [reply types/react-dom/index.d.ts]: I thought something with function overload. But other than that, simple comment should enough.

Comment (cynecx) [reply types/react-dom/index.d.ts]: I've pushed a new commit which adds a deprecated comment.

File: types/react-dom/test-utils/index.d.ts
Comment (vsiao) [inline types/react-dom/test-utils/index.d.ts]: What are the changes in this file for?

**Review by vsiao — COMMENTED** [review]
Please merge master to incorporate changes from https://github.com/DefinitelyTyped/DefinitelyTyped/pull/20065.

File: types/react-dom/test-utils/index.d.ts
Comment (cynecx) [reply types/react-dom/test-utils/index.d.ts]: To fix dtslint issues ;) The type parameter is only used once so the linter throws a stone at you.

Comment (johnnyreilly) [reply types/react-dom/test-utils/index.d.ts]: +1 for "throws a stone"

**Comment by cynecx** [comment]
Ping.

**Comment by circlingthesun** [comment]
^ @vsiao?

**Review by vsiao — APPROVED** [review]
Thanks!

**Comment by johnnyreilly** [comment]
Thanks all!

**Comment by plantain-00** [comment]
It seems it breaks a simple case:
```ts
import * as React from "react";
import * as ReactDOM from "react-dom";

class Main extends React.Component<{}, {}> {}

ReactDOM.render(<Main />, document.getElementById("container")); // <- error here
```

```bash
error TS2345: Argument of type 'Element' is not assignable to parameter of type 'ReactElement<any>[]'.
  Property 'length' is missing in type 'Element'.
```

**Comment by cynecx** [comment]
@plantain-00 I am not able to reproduce this. Your simple case compiles just fine here. What version of typescript are you currently using?

**Comment by plantain-00** [comment]
@cynecx 
```
"@types/react": "16.0.10",
"@types/react-dom": "16.0.1",
"typescript": "2.5.3",
```

**Comment by johnnyreilly** [comment]
@plantain-00 could you put together a minimal repro repo to share please? That will help things move quicker

**Comment by circlingthesun** [comment]
Based on @plantain-00's observation, I managed to work around this by adding the following to my package.json (I'm using yarn).
```
"resolutions": {
    "@types/react": "16.0.10",
    "@types/react-dom": "16.0.1",
}
```

**Comment by plantain-00** [comment]
it's OK now,  there is a `node_modules/@types/react-dom/node_modules/@types/react/*` in `node_modules/@types/react-dom`, reinstalling them works

**Comment by tkrotoff** [comment]
This does not work:
```TypeScript
  render() {
    return this.props.children;
  }
```

```
Types of property 'render' are incompatible.
  Type '() => ReactNode' is not assignable to type '() => string | number | false | Element | Element[] | null'.
    Type 'ReactNode' is not assignable to type 'string | number | false | Element | Element[] | null'.
      Type 'undefined' is not assignable to type 'string | number | false | Element | Element[] | null'.
```

Question: why do we have `ReactNode` and `JSX.Element`? Could they be merged?

**Comment by MartynasZilinskas** [comment]
@tkrotoff Since when we can return `undefined` in `render()`? You can only return `null` in `render` method. Unless it was changed in newest React v16  if children are absent it will return `null` instead of `undefined`.

**Comment by tkrotoff** [comment]
@MartynasZilinskas You're right, see https://codepen.io/tkrotoff/pen/yzKKdB

Still there is a problem:
```TypeScript
  render() {
    const { children } = this.props;
    return children !== undefined ? children : null;
  }
```

```
Types of property 'render' are incompatible.
  Type '() => string | number | boolean | {} | ReactElement<any> | (string | number | boolean | any[] | R...' is not assignable to type '() => string | number | false | Element | Element[] | null'.
    Type 'string | number | boolean | {} | ReactElement<any> | (string | number | boolean | any[] | ReactEl...' is not assignable to type 'string | number | false | Element | Element[] | null'.
      Type 'true' is not assignable to type 'string | number | false | Element | Element[] | null'.
```

**Comment by wcj3** [comment]
Has the above fix been added for stateless components? I'm getting an error when attempting to implement one that returns an array of elements

**Comment by patsissons** [comment]
why was the generic version of `findDOMNode` removed?
```ts
export function findDOMNode<E extends Element>(instance: ReactInstance): E;
```
(tagging @cynecx)

**Comment by vsiao** [comment]
@patsissons it was removed because of a lint failure. Does it suffice to convert usages like so?

before:
```ts
ReactDOM.findDOMNode<HTMLInputElement>(inst)
```

after:
```ts
<HTMLInputElement>ReactDOM.findDOMNode(inst)
```

**Comment by patsissons** [comment]
yea i am doing (basically) this as a temporary patch. what is the type safe way of finding a dom node of a specific type (i.e., my use case is that i am finding a dom node that will then be focused, so i need the type to be derived from `HTMLElement`)? I don't particularly enjoy using direct cast (or the `as` operator in my case, since it is in a tsx file).

**Comment by DovydasNavickas** [comment]
@patsissons, you will NOT get a type-safe result from a function `ReactDOM.findDOMNode`, because TypeScript's type system exists only during compile time and you are calling a function that was defined in JS world (`react-dom` library).

What you CAN do is have a type-gate for checking if an `element.focus != null && typeof element.focus === "function"`. This way you can create and use your own `Focusable` interface for that and have a type-safety this way.

**Comment by vsiao** [comment]
@patsissons I think the typecast is equivalent to what you were doing before. As @DovydasNavickas mentioned there's no guarantee that `findDOMNode` will return the exact subtype of `Element` that you expect.

**Comment by patsissons** [comment]
I have to imagine that the equivalence `findDOMNode<T>(elem)` is identical to `findDOMNode(elem) as T`. I was using the generic call with a semantic guarantee that it would return a sane dom node. But I now see that syntactically there ought to be no difference since the cast is just happening through generic parameterization rather than direct casting. Since i was already sanitizing the dom node anyways, it really doesn't matter other than requiring a direct cast which i try to avoid but in this case it is literally unavoidable (unless i create a type guard function to recast back to `HTMLElement`).

**Comment by niba** [comment]
@cynecx I think you forgot about functional components. The following code doesn't work
```
const Test: React.StatelessComponent<any> = () => [
  <div />,
  <div />,
]
```

and the error 

> [ts]
> Type '() => Element[]' is not assignable to type 'StatelessComponent<any>'.
>   Type 'Element[]' is not assignable to type 'ReactElement<any> | null'.
>     Type 'Element[]' is not assignable to type 'ReactElement<any>'.
>       Property 'type' is missing in type 'Element[]'.

**Comment by cynecx** [comment]
@niba Seems so. Looks like a trivial change to me. I can open a PR later.

**Comment by niba** [comment]
@cynecx I have been trying to add this on my own but it doesn't look like a trivial change. Simple changing the return type of StatalessComponent function breaks JSX
from` (props: P & { children?: ReactNode }, context?: any): ReactElement<any> | null;`
to  `(props: P & { children?: ReactNode }, context?: any): ReactElement<any>[] | ReactElement<any>| null;`

with this change, the below definition of functional component started working
```
const StatelessComponent: React.SFC<SCProps> = ({ foo }: SCProps) => {
  return [
    <div>a</div>,
    <div> aa</div>
  ];
};
```
but now it's impossible to call it from JSX. `<StatelessComponent />;` produces

> [ts]
> JSX element type 'ReactElement<any> | ReactElement<any>[] | null' is not a constructor function for JSX elements.
>   Type 'null' is not assignable to type 'ElementClass'.
> [ts] JSX element class does not support attributes because it does not have a 'props' property.

I don't see any constructor function in ReactElement and I don't know how JSX is processed by typescript. Do you have any idea what is wrong?

**Comment by cynecx** [comment]
@niba You are right, this doesn't looks trivial. I am also not sure what's wrong here (I am not too familiar with this code) but I will try investigating this.

**Comment by vsiao** [comment]
I don't have much time to look into this, but there's some useful background in https://github.com/Microsoft/TypeScript/issues/5478 about how SFC+TSX works. cc @RyanCavanaugh

**Comment by notitatall** [comment]
I get an error when I try to return an array that contains a null element from `render()`. Simplified example:

```
class ReturnArrayWithNullElement extends React.Component<Props, {}> {

    ...

    render() {
        return [
            <div key={1}>Element 1</div>,
            <div key={2}>Element 2</div>,
            this.props.displayThird ? <div key={3}>Element 3</div> : null
        ]
    }
}
```

Maybe this isn't an idiomatic way of conditionally rendering an element inside of an array (still becoming acquainted with the Fiber), but React deals with this how I would expect. TypeScript throws the following error:

```
error TS2415: Class 'ReturnArrayWithNullElement' incorrectly extends base class 'Component<Props, {}>'.
  Types of property 'render' are incompatible.
    Type '() => (Element | null)[]' is not assignable to type '() => string | number | false | Element | Element[] | ReactPortal | null'.
      Type '(Element | null)[]' is not assignable to type 'string | number | false | Element | Element[] | ReactPortal | null'.
        Type '(Element | null)[]' is not assignable to type 'ReactPortal'.
```

Any thoughts?

**Comment by vsiao** [comment]
@notitatall the return type of `render` needs to be merged with the `ReactNode` type. For the time being you can probably cast to an appropriate type to unblock yourself.

**Pull request merged**
